File: Easy/345_Reverse_Vowels_of_a_String/Solution.java


public class Solution {
    Boolean isVowel(char c) {
        c = Character.toLowerCase(c);
        return c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u';
    }
    
    String reverseVowels(String s) {
        char[] chars = s.toCharArray();
        int left = 0;
        int right = s.length() - 1;

        while (left < right) {
            while (left < right && !isVowel(chars[left])) {
                left++;
            }
            while (left < right && !isVowel(chars[right])) {
                right--;
            }

            if (left < right) {
                char temp = chars[left];
                chars[left] = chars[right];
                chars[right] = temp;

                left++;
                right--;
            }
        }

        return new String(chars);
    }

    public static void main(String[] args) {
        Solution solution = new Solution();

        // test cases 1
        System.out.println(solution.reverseVowels("IceCreAm"));

        // test cases 2
        System.out.println(solution.reverseVowels("leetcode"));
        
    }
}

File: Easy/345_Reverse_Vowels_of_a_String/Solution.py
from typing import List, Optional


class Solution():
    def reverseVowels(self, s):
        vowels = set("aeiouAEIOU")
        l = [char for char in s if char in vowels]
        result = []

        for char in s:
            if char in vowels:
                result.append(l.pop())
            else:
                result.append(char)

        return "".join(result)


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    print(solution.reverseVowels("IceCreAm"))

    # test cases 2
    print(solution.reverseVowels("leetcode"))
    